File: src/Longest Subsequence with Adjacent Diff as 1/Code/solution.java
class Solution {
    public int longestSubseq(int[] arr) {
        // dp[x] stores the longest valid subsequence ending with value x.
        HashMap<Integer, Integer> dp = new HashMap<>();

        // This stores the longest subsequence found so far.
        int ans = 0;

        // Process the array from left to right to preserve subsequence order.
        for (int x : arr) {
            // Get the best subsequence ending at x - 1, or 0 if it does not exist.
            int left = dp.getOrDefault(x - 1, 0);

            // Get the best subsequence ending at x + 1, or 0 if it does not exist.
            int right = dp.getOrDefault(x + 1, 0);

            // Extend the better neighboring subsequence with the current x.
            int current = Math.max(left, right) + 1;

            // Store the best length ending with value x.
            dp.put(x, current);

            // Update the overall maximum length.
            ans = Math.max(ans, current);
        }

        // Return the longest valid subsequence length.
        return ans;
    }
}

File: src/Longest Subsequence with Adjacent Diff as 1/Code/solution.cpp
class Solution
{
public:
    int longestSubseq(vector<int> &arr)
    {
        // dp[x] stores the longest valid subsequence ending with value x.
        unordered_map<int, int> dp;

        // This stores the longest subsequence found anywhere in the array.
        int ans = 0;

        // Process elements from left to right so only previous elements are used.
        for (int x : arr)
        {
            // A valid previous value must be exactly x - 1 or x + 1.
            int left = dp.count(x - 1) ? dp[x - 1] : 0;
            int right = dp.count(x + 1) ? dp[x + 1] : 0;

            // Extend the longer subsequence by adding the current element x.
            int current = max(left, right) + 1;

            // Store the best length ending with x.
            dp[x] = current;

            // Update the overall maximum answer.
            ans = max(ans, current);
        }

        // Return the length of the longest valid subsequence.
        return ans;
    }
};